### ISSUE TITLE
Calculating conversion rates other than T>C

### ISSUE THREAD

**Issue by goeringr** [t-neumann/slamdunk#97]
Hello,

I am interested in detecting an RNA modification that is read as G>C or G>T.
I have been using the slamdunk pipeline but ignoring the final tCount step. 
I can detect my desired signal using the alleyoop utrrates function however, I wanted to check that this is analogous to calculating T conversions and T coverage in the tCount function. 

Is the T_T (or G_G) column calculated by alleyoop utrrates equivalent to coverage of reads on these bases for each UTR?

Using only utrrates, I can calculate a summed conversion rate of (G_C + G_T)/#ReadsonGs(G_G) for each UTR.

Is this an accurate way of calculating conversion rates other than T>C?

Thanks for your insight
--Rae

**Comment by t-neumann** [comment]
Hi @goeringr,

first of all:

Is the T_T (or G_G) column calculated by alleyoop utrrates equivalent to coverage of reads on these bases for each UTR?
- yes that is correct

Is this an accurate way of calculating conversion rates other than T>C?
Within the Slamdunk framework not, for this you would write a custom script but of course you could orient yourself on the tcounting script.

What you have to bear in mind however is that if you use the T>C conversion aware scoring that Slamdunk provides, you would bias your dataset towards T>C mismatches. So you could also run the map step in `quantseq` mode to deactivate this scoring and have a more accurate representation of your mismatch rates.

Tobi

**Comment by goeringr** [comment]
Hello again Tobi,

I've been using alleyoop utrrates to calculate conversion rates other than T>C and first tried to replicate the data created by tcount() with alleyoop utrrates.

I am wondering why different Coverage and Conversions are calculated from the same exact data and same UTR bed file with these two different functions. Most UTRs are treated exactly the same for each method but some UTRs behave very differently...

![image](https://user-images.githubusercontent.com/29406405/127568181-334af54a-1059-4e3e-9407-0c04c427f83f.png)
![image](https://user-images.githubusercontent.com/29406405/127568417-4bb09632-b552-4043-b7c6-ebe0f603f090.png)

I was hoping you could explain the difference between these two functions in calling conversions and why tcount seems to count more T>C conversions than utrrates. 

thank you
--Rae

**Comment by t-neumann** [comment]
Hi Rae,

hm haven't looked into the `utrrates` function for a while - it seems there it ignores any reads that are not carrying a valid T>C conversions which is on the other hand taken into account for by the `tcount` function:

https://github.com/t-neumann/slamdunk/blob/cf9bbbdebcb86ca1cc0c60aa9ea268deb565cd25/slamdunk/dunks/stats.py#L368

Could this explain the difference?

**Comment by taliaferrojm** [comment]
Hi Tobi,

Thanks so much for your help.  I'm slightly confused as to how the above statement could ever evaluate to True.  How could a read both not be a TcRead and also have a tcCount > 0?

Regardless, it seems like you are telling us that the best thing for us to do is rework the `tcount` code to suit our purposes (counting conversions of Gs to either C or T).  This is something we can probably handle, but I want to make sure we don't miss also modifying any code that `tcount` calls or considerations it makes that are T-> C specific (the SNP evaluator, any alignment considerations, etc.). Would it be possible to just get a list of functions or tools off the top of your head that we should be sure to check?  For an outsider, the code is a bit complicated, and if we knew places we should have a look at, that would be a huge help.

Thanks!
Matt

**Comment by t-neumann** [comment]
Hi Matt,

yeah that is kind of a hack I introduced back then when we wanted to have a threshold for a minimum number of `T>C` conversions to have a read be counted as `tc-read`.

Currently the `tcount` function uses proprietary BAM-tags produced by NextGenMap to very fast and efficiently extract `T>C` conversions. What we did previously was to use `pysam` and extract them ourselves by iterating over reads in a given interval and this you could also do with any type of base conversion.

For alignments, depending on your base conversion rate, you might be able to get away with simply using common aligners like STAR and the SNP function simply calls Varscan2, so you can also just prefilter those results for any other base-conversion.

So in summary, the functions I would use and modify are the `map` function, to use either NextGenMap with a neutral scoring scheme (not T>C aware) or STAR. Dependent on that you can possibly skip the `filter` function for STAR.
`snp` should in principle work, since it just produces a vcf file with all possible variants.

And then finally `tcount` you would need to extract your desired base-conversions yourself from the read set, an easy way to do is is `pysam`.

Best,

Tobi
